### ISSUE TITLE
Docker Engine Memory leak

### ISSUE THREAD

**Issue by andresnatanael** [docker/for-mac#178]
### Expected behavior

The process com.docker.hyperkit after killing all the docker containers need to free the used memory to the initial state (50 mb)

### Actual behavior

After killing all the docker containers the process com.docker.hyperkit still using 3.49 GB.

### Information

Diagnostic ID: EB6AFE2E-34AA-4617-B849-D79863CDC40C
Docker for Mac: 1.12.0 (Build 10871)
macOS: Version 10.11.6 (Build 15G31)
[OK]     docker-cli
[OK]     app
[OK]     moby-syslog
[OK]     disk
[OK]     virtualization
[OK]     system
[OK]     menubar
[OK]     osxfs
[OK]     db
[OK]     slirp
[OK]     moby-console
[OK]     logs
[OK]     vmnetd
[OK]     env
[OK]     moby
[OK]     driver.amd64-linux

### Steps to reproduce

1. ...
docker-compose up -d

version: '2'
services:
  student:
    image: docker:dind
    ports:
      - "8000-8010"
    privileged: true
    volumes:
      - /tmp/docker-training:/docker-training

2. ...

docker-compose scale student=10

3.

docker-compose down

4.-

docker ps -> no containers running, but consuming a lot of memory com.docker.hyperkit (3,49 gb)

Only restarting the docker engine VirtualMachine (using hyperkit) free the memory.

**Comment by ijc25** [comment]
Thanks for your report. Unfortunately once the memory has been touched by the Linux kernel within the VM and therefore becomes populated RAM in the hyperkit process (by the usual OS page faulting mechanisms) there is no way for the guest kernel to then indicate back to the hypervisor and therefore the host when that RAM is free again and to turn those memory regions back into unallocated holes.

However since the RAM is unused in the guest it should not be touched by anything in the VM and therefore not by the hyperkit process either and therefore I would expect it to eventually get swapped out to disk in favour of keeping actual useful/active data for other processes in RAM, just as it would for any large but idle process.

It sounds like you are checking the virtual address size (`vsz` in `ps` output) of the hyperkit process rather than the resident set size (`rss`). The later should be shrinking for an idle hyperkit process as other processes request memory and hyperkit gets swapped out while the former basically only grows but does not necessarily represent use of actual RAM.

I'm afraid that compacting the `vsz` is basically a wont-fix issue here and I'm therefore going to close on that basis. If however you are observing the `rss` not shrinking (and there are other processes to create memory pressure i.e. the memory appears to be somehow locked into RAM and not swappable) then please do update this ticket with details of the `rss` memory patterns you are observing and we can reopen and investigate that angle further.

**Comment by nmdias** [comment]
Couldn't we open a bug report to Apple to improve upon the Hypervisor.framework? 

I attempted to run Veertu Desktop which also runs on the Hypervisor.framework, with a memory limit of 1GB and it didn't use more then 900MB. With docker, just pulling mysql vanilla with no additional databases running it sucks up about 1,78GB of memory.

**Comment by eXamadeus** [comment]
Hey all.  Sorry to revive a dead thread, but I would like this ticket to be reopened.  After looking into it, the reserved memory for hyperkit process sits around 1.6-1.7 GB even when trying to apply significant memory pressure.  (This is of course without any running containers)

Output of `docker info`:

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 56
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.87-linuxkit-aufs
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 3
Total Memory: 5.818GiB
Name: linuxkit-025000000001
ID: ZMMF:UZTZ:IHNE:2EIZ:TA4G:6P4L:SU3N:4OSK:UC5K:UM4M:DIKT:6ZXJ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 22
 Goroutines: 41
 System Time: 2018-06-29T18:06:39.2425914Z
 EventsListeners: 2
HTTP Proxy: docker.for.mac.http.internal:3128
HTTPS Proxy: docker.for.mac.http.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

I wrote a simple python script to store crap into memory, then I spam opened as many google chrome tabs as I could (best way to apply memory pressure I could think of).  Here is the script I used:

```python
import sys
import threading
from unittest import TestCase

items = []

class WastefulThread(threading.Thread):

    def run(self):

        for i in range(0, 10000000000):
            items.append(i)
            if i % 100000 == 0:
                sys.stdout.write('\r' + str(i))
                blah = items[0:i]

        for item in items:
            sys.stdout.write('\r{}'.format(item))


class WasteMemory(TestCase):

    def test_waste(self):
        for i in range(8):
            thread = WastefulThread(name='Thread-{}'.format(i))
            thread.start()
```

and here is a gif showing the memory usage of the top four processes sorted by reserved memory (RES):

![memory](https://thumbs.gfycat.com/OrderlyEvenKawala-size_restricted.gif)

The memory of footprint of the `com.docker.hyperkit` process stays constant at >1600 MB.  This is the RES footprint.  Also, note that there is SIGNIFICANT memory pressure applied to the system, so much so that the current running "WasteMemory" process ends up dumping its memory to virtual (VIRT) quite a few times throughout the gif.

Here is a [higher quality gifv link](https://gfycat.com/OrderlyEvenKawala).

TLDR: It appears that the `com.docker.hyperkit` process does not free up reserved memory, even if significant memory pressure is applied.  I would like this issue to be reopened and investigated as this is a significant amount of memory being reserved for what should be an inactive process.

**Comment by eXamadeus** [comment]
I have also crawled around the related issues and haven't found anything that explains the reasons or points to a solution for this issue.  Please let me know if I can provide any more information.  I believe the GIF I provided doesn't show full memory reservation (roughly 16 GB); however, during other tests I ran it was almost 100% maxed through similar means and the RES portion of the `com.docker.hyperkit` process remained unchanged.

**Comment by mkohlmyr** [comment]
Having docker idle in my toolbar lands it at 1.3GB of RAM I really don't understand how it is even possible. This is very shortly after booting, before using any containers.

**Comment by iddan** [comment]
Can this be reopened please?

**Comment by kyprifog** [comment]
I just opened my computer with no containers running and found hyperkit using 9GB of RAM.  This needs to be reopened.

**Comment by gvbkr** [comment]
Yes  please.. with no containers open, its ~5G of RAM being used. This is significantly impacting the avaialble resources and slowing down other things we need to run. A possible bug introduced in the latest version?

**Comment by AnnaKarinaNava** [comment]
same situation here, just restarted my laptop, no containers and  com.docker.hyperkit ~2G of RAM

**Comment by vfontes** [comment]
Exact same situation here as well. Sitting at 2.61 GB without a single container running.

**Comment by ICT22** [comment]
2.91 GB here. top of the chart for me :) Just doesn't seem right.

**Comment by Mibeon** [comment]
Something is going wrong.
I read all the comments and realized that also in my case the usage of about 1.4 GB (for nothing) is too much. I asked me lots of time why does docker need so much ram at blank running (event with no containers created).

Reopen this ticket, damn. The explanation at the top is not plausible enough.

**Comment by iMerica** [comment]
I'm seeing the same thing on `Docker version 18.06.1-ce, build e68fc7a`. 

PSA: If you're reporting the same issue, please attach the version of Docker you're running using `docker --version`. Its just useless noise otherwise.

**Comment by penndsg** [comment]
Same problem, `Docker version 18.06.0-ce, build 0ffa825`, 1.8 GB with no containers running.

**Comment by ICT22** [comment]
I'm running at 5.05Gb now, brand new MacBook never even run a docker image on it. This can't be right.

`Docker version 18.06.1-ce, build e68fc7a`

`Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.93-linuxkit-aufs
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 11.71GiB
Name: linuxkit-025000000001
ID: ARNV:QEQI:MLN4:XLZ6:PFY7:44QK:XGEQ:GX6H:RZTY:6JAJ:CBBI:ZS6R
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 24
 Goroutines: 50
 System Time: 2018-09-27T06:46:47.768576607Z
 EventsListeners: 2
HTTP Proxy: gateway.docker.internal:3128
HTTPS Proxy: gateway.docker.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
`

**Comment by stroebs** [comment]
Same issue here on a fresh installation of MacOS Mojave. Docker-CE 18.06.1-ce-mac73 (26764)

![image](https://user-images.githubusercontent.com/1121790/46145190-ac1c0800-c25f-11e8-82d9-0b8db635e4da.png)
![image](https://user-images.githubusercontent.com/1121790/46145192-ade5cb80-c25f-11e8-93fb-a88483456035.png)
![image](https://user-images.githubusercontent.com/1121790/46145256-ea192c00-c25f-11e8-8bcb-747061e1cb30.png)

**Comment by fadelardiSA** [comment]
2 GB usage here. No containers running. 
Docker for Mac ver.: Version 18.06.1-ce-mac73 
MacOS: 10.12.6

**Comment by troyharvey** [comment]
🤔  no containers running

![image](https://user-images.githubusercontent.com/530530/46159467-6276f180-c24e-11e8-9b15-bcb874678863.png)

**Comment by daBayrus** [comment]
no containers running.. OS: **Mojave** Version **18.06.1-ce-mac73 (26764)**
![image](https://user-images.githubusercontent.com/751749/46188550-5bb6b000-c31c-11e8-8cbf-673da46b0927.png)

**Comment by EduardRakov** [comment]
cpu is loaded too much as well as ram and my macbook lagging :( this problem was appeared after updating from high sierra to mojave. It's really annoying, folks, do you have maybe some thoughts/estimates when the problem can be fixed?

**Comment by andresnatanael** [comment]
The problem I thing is related on how is virtualized docker and native macOS Virtualization using the framework xhyve. Is there anyway to improve that or avoid virtualization? Someone of the developer can re-open this? Too many comments...

**Comment by fr0der1c** [comment]
`Docker version 18.06.1-ce, build e68fc7a`
Hyperkit is taking 2.79GB memory with no container running even if I limit the memory available for Docker to 2GB. Restarting Docker won't help.

**Comment by eddiemonge** [comment]
The original issue was never solved and thus shouldn't have been closed. Opening new issues just creates duplicate issues and makes searching for resolutions that much more difficult and time consuming

**Comment by tryggvigy** [comment]
Mac `High Sierra 10.13.6 (17G65)`
Docker `Version 18.06.1-ce-mac73 (26764)` 
Completely fresh install. No containers or images even exist on my machine.

![image](https://user-images.githubusercontent.com/2373958/47160090-79f64700-d2ef-11e8-8af3-7aaed2b03275.png)

**Comment by scripterkaran** [comment]
Memory:

<img width="573" alt="screen shot 2018-10-20 at 1 47 54 am" src="https://user-images.githubusercontent.com/10154253/47241687-3cd2a780-d40a-11e8-8a9a-bd605eacc70b.png">

OS: 
<img width="254" alt="screen shot 2018-10-20 at 1 48 54 am" src="https://user-images.githubusercontent.com/10154253/47241751-71defa00-d40a-11e8-9c9e-ecf294527157.png">


Docker:
<img width="313" alt="screen shot 2018-10-20 at 1 49 19 am" src="https://user-images.githubusercontent.com/10154253/47241764-7d322580-d40a-11e8-8603-3757ecfb3804.png">

**Comment by perseghini** [comment]
Same issue here.
And stopping Docker makes my laptop runs much better (no more lagging and slowness in MacOS).
And I don't have any container running.
Only solution is to stop Docker and only run it when I need to use it.
MacOS: 10.13.6
![screen shot 2018-10-23 at 11 35 27 am](https://user-images.githubusercontent.com/4202929/47382884-3701ee00-d6b8-11e8-9756-5874a1418e08.png)

**Comment by kreintjes** [comment]
Same issue over here. With no containers running hyperkit uses 2.7 GB already. When I start the two containers I need, it uses 3+ GB RAM. I'm running MacOS 10.14 (18A391) and Docker Version 18.06.1-ce-mac73 (26764).

limiting the amount of resources available to docker (Preferences -> Advanced) to 1 GB Memory and 512 MB Swap helps a bit. Hyperkit now uses 1.6 GB without any containers running and 1.8 GB with the two containers running. The containers still perform well enough.

**Comment by mcgitty** [comment]
Same issue, 1.31GB casualty with 0 container, macOS 10.12.6 (Siera), Docker 18.06.1-ce-mac73 (26764).

**Comment by AbrahamAlcaina** [comment]
macOS 10.14.1 / 18.06.1-ce-mac73 (26764)
![image](https://user-images.githubusercontent.com/3775437/47842207-4cfc6680-ddbc-11e8-9535-60681d027da7.png)

**Comment by matt9949** [comment]
macOS 10.14 (18A391):

<img width="800" alt="screenshot 2018-11-01 at 11 37 33" src="https://user-images.githubusercontent.com/7140752/47849846-f5192c00-ddca-11e8-8903-504144654be1.png">

No containers running:

![screenshot 2018-11-01 at 11 38 01](https://user-images.githubusercontent.com/7140752/47849853-fb0f0d00-ddca-11e8-9bbe-46fa7c5f7124.png)

**Comment by mnorkin** [comment]
MacOS Mojave, running 0 containers

![image](https://user-images.githubusercontent.com/55679/47918774-30445980-deb6-11e8-9e4f-2616fffb17f4.png)

**Comment by naviat** [comment]
MacOS Mojave,  0 container
<img width="474" alt="screen shot 2018-11-06 at 17 01 38" src="https://user-images.githubusercontent.com/13325589/48057058-b1b52800-e1e5-11e8-855e-abb141181f5f.png">

**Comment by mcgitty** [comment]
We can stop complaining. The first reply to this thread "ijc commented on Aug 9, 2016" stated that hogging the memory is a hyperkit VM behavior:

https://docs.docker.com/docker-for-mac/docker-toolbox/#the-docker-for-mac-environment
https://github.com/moby/hyperkit
https://developer.apple.com/documentation/hypervisor

On Linux, there is no such problem, because the host kernel supports cgroup. Not all Docker engines are created equal.

**Comment by pmingram** [comment]
> We can stop complaining. The first reply to this thread "ijc commented on Aug 9, 2016" stated that hogging the memory is a hyperkit VM behavior:

Without people reporting an obvious major flaw in a system, we can never progress and improve.

**Comment by pabloleone** [comment]
I'm moving to docker because it takes fewer resources...jajaja

It's taking 4GB of RAM to run with no containers started. I understand the reply, but still an issue. Any solutions?

**Comment by iMerica** [comment]
@mcgitty 

> We can stop complaining.

A Github issue is not a "Complaint" . 

https://en.wikipedia.org/wiki/Software_bug

**Comment by envygeeks** [comment]
> On Linux, there is no such problem, because the host kernel supports cgroup.

So because there is no bug on Linux, that means there must be no bug on MacOS? This is a very real problem.  And the original explanation of why this is happening is complete hogwash.  I've only booted up Alpine Linux, you're telling me that Alpine Linux running on Docker for MacOS used 3gb of ram, and now suddenly Docker can't reclaim it through Hyperkit?

**Comment by ahayes99** [comment]
<img width="426" alt="screenshot 2018-11-15 at 15 15 36" src="https://user-images.githubusercontent.com/44228255/48562071-57088400-e8e9-11e8-9da0-00ff99148dd9.png">

up to 10gb now.. keep up the good work docker

**Comment by gvbkr** [comment]
At this point, its not practical to run any docker containers in my 2018 MAC as a basic django app running inside docker container hogs upto 28G of Memory on the MAC. This is a real problem and it forced me to just use virtualenv on the MAC directly instead of using the container.

Does anyone who works on this repo uses MAC for development? I am assuming many engineers working at real companies are facing this problem, so we need a solution to solve this problem.

**Comment by robotdan** [comment]
Same here, I fire up Docker so I can build and publish images, and then I have to keep it turned off otherwise it chews through all of my RAM. 

The offending process is always `com.docker.hyperkit`

MacBook Pro (15-inch 2016)
16 GB RAM
macOS Mojave version 10.14.1

**Comment by iMerica** [comment]
It would be nice if someone from Docker would comment on this issue to let us know either A) They don't have the bandwidth to fix this bug or B) They are aware of the issue and someone is already working on it. 

The lack of communication is more frustrating than the bug itself.

**Comment by KatSick** [comment]
Same for me. 1 container running
<img width="807" alt="image" src="https://user-images.githubusercontent.com/13982649/48804289-eb754b00-ed1c-11e8-872d-382a6979b6ba.png">

**Comment by robotdan** [comment]
Thanks @KatSick same here. 

@ijc Since this issue is closed, can you advise on how we can best move forward here? 

I think can all agree that an idle process should not consumer 4+ GB of RAM without ever releasing it. There certainly does seem to be some sort of leak going on here. 

Could we collect additional debug information to assist in resolving this issue, or based upon your knowledge of the system, do you have any guesses on where this issue may reside so we can open bugs elsewhere?

From the end user perspective, the `com.docker.hyperkit` consumes a lot of RAM when it is idle so this does seem to a reasonable place to open issues, but if you can please provide some insight into where we should focus our attention that would be really helpful. 

Thanks, I'm happy to help in any way that I can.

**Comment by eddiemonge** [comment]
Another issue is open for the same thing  #3232

**Comment by iMerica** [comment]
This issue persists in the newest version just released (`2.0.0.0-mac78 28905`). 

```
Release Notes:
* Upgrades
  - [Docker 18.09.0](https://github.com/docker/docker-ce-packaging/releases/tag/v18.09.0)
  - [Docker compose 1.23.1](https://github.com/docker/compose/releases/tag/1.23.1)
  - [Docker Machine 0.16.0](https://github.com/docker/machine/releases/tag/v0.16.0)
  - [Kitematic 0.17.5](https://github.com/docker/kitematic/releases/tag/v0.17.5)
  - Linux Kernel 4.9.125

* New
  - New version scheme

* Deprecation
  - Removed support of AUFS
  - Removed support of OSX 10.11

* Bug fixes and minor changes
  - Fix appearance in dark mode for OSX 10.14 (Mojave)
  - VPNKit: Improved scalability of port forwarding. Related to [docker/for-mac#2841](https://github.com/docker/for-mac/issues/2841)
  - VPNKit: Limit the size of the UDP NAT table. This ensures port forwarding and regular TCP traffic continue even when running very chatty UDP protocols.
  - Ensure Kubernetes can be installed when using a non-default internal IP subnet.
  - Fix panic in diagnose
```

**Comment by adampatterson** [comment]
My Docker Deamon was using just over 4 gigs of memory without anything running.

Containers: 76
 Running: 0
 Paused: 0
 Stopped: 76
Images: 446
Server Version: 18.09.0
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.125-linuxkit
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: UDXG:4JIP:BRDP:D4G6:PPX7:TVA7:IHRY:5WWZ:S77K:L6M3:MV6I:5L3G
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 24
 Goroutines: 50
 System Time: 2018-11-22T07:39:47.055602708Z
 EventsListeners: 2
HTTP Proxy: gateway.docker.internal:3128
HTTPS Proxy: gateway.docker.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

**Comment by eexbee** [comment]
Same issue on Mojave

**Comment by azamuddin** [comment]
Same issue on mac Mojave, eating 4GB of my RAM with no containers running

**Comment by andrewalf** [comment]
Please fix memory and cpu usage, empty docker eats ram and battery. Macos Mojave too :(

**Comment by luisfdeandrade** [comment]
Same issue here! [
![screen shot 2018-11-28 at 11 28 44](https://user-images.githubusercontent.com/3513085/49154926-d55d2100-f300-11e8-9353-aa5f3fb32b3e.png)
](url)

**Comment by mcgitty** [comment]
> We can stop complaining. The first reply to this thread "ijc commented on Aug 9, 2016" stated that hogging the memory is a hyperkit VM behavior:
> 
I meant we may be asking the wrong team to fix this. Probably this one: https://github.com/moby/hyperkit

That said, the docker for Mac team doesn't seem to care enough to reply here and promise to work with the hyperkit team.

**Comment by robotdan** [comment]
This issue looks promising on the hyperkit project. 
https://github.com/moby/hyperkit/issues/231

**Comment by KyeRussell** [comment]
@mcgitty I'm sure they'll give you your money back. 🙄

Please keep things professional.

**Comment by mcgitty** [comment]
@KyeRussell You mean they'll give me my "memory" back. I certainly hope so, soon.

I just want to remind everyone, this ticket is "Closed", so unlikely to get any attention from the team. We should follow the moby/hyperkit#231 issue pointed out by robotdan.

**Comment by gaf3** [comment]
Ya I'm at 3G of RAM with nothing running after restarting Docker.  This is crazy.  It's using nearly 1/4 of my resources and not doing anything. I wouldn't care except my Mac is swapping and becomes unusable for a few minutes.

**Comment by robabby** [comment]
I installed Mojave this weekend but I experienced no issues yesterday.  Today suddenly Docker seems to be a huge memory hog 🤔 

<img width="803" alt="screen shot 2018-12-18 at 10 23 35 am" src="https://user-images.githubusercontent.com/736269/50167558-0bae1080-02af-11e9-96e0-7b8e4d454e16.png">

**Comment by adampatterson** [comment]
Same, I am at 4 gigs I don't have a VM running.  But I have had this issue for months.

Even after the update.

**Comment by joekohlsdorf** [comment]
This is a bug in macOS, double click on the process to open its details, there you'll see the real memory usage.
Virtual memory isn't allocated.

<img width="527" alt="screen shot 2018-12-28 at 12 11 39" src="https://user-images.githubusercontent.com/32424163/50519547-2eec7300-0a9a-11e9-801e-0c88de1d78a2.png">

**Comment by mcgitty** [comment]
> This is a bug in macOS, double click on the process to open its details, there you'll see the real memory usage.
> Virtual memory isn't allocated.
> 
Joe, you can add the 'Real Memory' column to the main Activity Monitor view. I don't know how you get that fantastic number. My macOS Sierra + Docker 18.06.1 with nothing running looks like this:

<img width="705" alt="screen shot 2018-12-28 at 9 47 27 am" src="https://user-images.githubusercontent.com/4326639/50523724-8da7f180-0a86-11e9-908f-f4f384e746a4.png">

**Comment by jcouso** [comment]
Same issue, I'm seeing 1,32GB of real memory being used by docker with no containers running. I'm using Mojave with Docker 2.0.0. Is this issue every going to get address?

![docker2](https://user-images.githubusercontent.com/28370865/50684259-b7c05d00-100c-11e9-8458-6102ae1e808f.png)

```
Containers: 17
 Running: 0
 Paused: 0
 Stopped: 17
Images: 9
Server Version: 18.09.0
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.125-linuxkit
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: 2UVK:FUYF:2QVF:UAT5:42N7:5CZO:6TPJ:HMW2:AMIU:5XU4:QU3N:U7IZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 24
 Goroutines: 50
 System Time: 2019-01-04T10:34:21.087308216Z
 EventsListeners: 2
HTTP Proxy: gateway.docker.internal:3128
HTTPS Proxy: gateway.docker.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Comment by alexgurr** [comment]
> This is a bug in macOS, double click on the process to open its details, there you'll see the real memory usage.
> Virtual memory isn't allocated.
> 
> <img alt="screen shot 2018-12-28 at 12 11 39" width="527" src="https://user-images.githubusercontent.com/32424163/50519547-2eec7300-0a9a-11e9-801e-0c88de1d78a2.png">

I'm seeing something similar to this. Initially looked like really high memory usage, but in fact I think this is just pre-allocated and is not in fact used.

https://puu.sh/CsO1g/85a311b7e7.png

**Comment by kamil-jakubowski** [comment]
#3232 
#3304

**Comment by Lukasz-pluszczewski** [comment]
@alexgurr Actually, the Real Memory Size is smaller but still way too big (no containers running):
<img width="560" alt="Screenshot 2019-05-17 at 16 00 51" src="https://user-images.githubusercontent.com/19817661/57933270-1f40c000-78bd-11e9-9e64-05c5e9fecb79.png">

**Comment by javiermolinaalonso** [comment]
Same issue here. Fresh MacOS Mojave installation.

> > docker info
> Containers: 2
>  Running: 2
>  Paused: 0
>  Stopped: 0
> Images: 2
> Server Version: 18.09.2
> Storage Driver: overlay2
>  Backing Filesystem: extfs
>  Supports d_type: true
>  Native Overlay Diff: true
> Logging Driver: json-file
> Cgroup Driver: cgroupfs
> Plugins:
>  Volume: local
>  Network: bridge host macvlan null overlay
>  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
> Swarm: inactive
> Runtimes: runc
> Default Runtime: runc
> Init Binary: docker-init
> containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
> runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
> init version: fec3683
> Security Options:
>  seccomp
>   Profile: default
> Kernel Version: 4.9.125-linuxkit
> Operating System: Docker for Mac
> OSType: linux
> Architecture: x86_64
> CPUs: 2
> Total Memory: 1.952GiB
> Name: linuxkit-025000000001
> ID: JDRP:AT3A:GGVA:5D65:DLBZ:TOW2:2CJE:37KZ:OAPH:KTXZ:YK73:CZIL
> Docker Root Dir: /var/lib/docker
> Debug Mode (client): false
> Debug Mode (server): true
>  File Descriptors: 24
>  Goroutines: 51
>  System Time: 2019-05-20T07:21:42.3099723Z
>  EventsListeners: 2
> HTTP Proxy: gateway.docker.internal:3128
> HTTPS Proxy: gateway.docker.internal:3129
> Registry: https://index.docker.io/v1/
> Labels:
> Experimental: false
> Insecure Registries:
>  127.0.0.0/8
> Live Restore Enabled: false
> Product License: Community Engine

![Screenshot 2019-05-20 at 09 26 02](https://user-images.githubusercontent.com/4775160/58003626-71622b00-7ae1-11e9-85e4-b213e13b8ad4.png)

**Comment by katafractari** [comment]
Just installed Docker for Mac. `com.docker.hyperkit`'s "real memory size" is 1,5 GB before starting any containers.

![image](https://user-images.githubusercontent.com/1299228/58583044-6763c900-8253-11e9-9109-03f34035cd12.png)

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.125-linuxkit
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: J2EE:T2T4:6UIF:UGIT:TSQ6:XP2A:SAXK:E45U:3V6G:XSKB:UU5I:XJ6R
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 24
 Goroutines: 50
 System Time: 2019-05-29T18:50:49.644643833Z
 EventsListeners: 2
HTTP Proxy: gateway.docker.internal:3128
HTTPS Proxy: gateway.docker.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Comment by adampatterson** [comment]
I don't see any reports that report consistent memory usage.

![image](https://user-images.githubusercontent.com/31537/58586233-d58a9a80-8217-11e9-8681-b341d0df64eb.png)

**Comment by OItsMAK** [comment]
Same problem here.
I'm running nothing in docker (literally); Yet see a Memory usage of 4.05 GB.

And when the docker daemon is running my system (an iMac with 16GB of memory, with ~3GB of it free) is oddly choked and everything gets slow. and when I stop the Docker daemon it all gets smooth.

Also I've tried removing redundant and unimportant local images and stopped containers (removed like 40 old-version-tagged images!) and the memory usage reduced to 3.94 GB! Even after I restart the daemon.

**Comment by ianldgs** [comment]
I didn't want update from High Sierra to Mojave because I was afraid something like that would happen. A few days ago my MacBook got updated after being sent to apple care, and now I'm having the same problem.

**Comment by MatayoshiMariano** [comment]
Is there any workaround? Experiencing the same thing. This is very frustrating 😬

**Comment by stevetbright** [comment]
I am running on a mac with macOS 10.14.5

With no containers running after a reboot (?!) I see 3 GB of memory usage

Is docker doing crypto mining or something in the background?

<img width="753" alt="Screen Shot 2019-06-07 at 9 47 51 AM" src="https://user-images.githubusercontent.com/5326593/59120146-7dcfeb80-8909-11e9-9ac4-47af24e76b83.png">

**Comment by MatayoshiMariano** [comment]
Just saw [this](https://forums.docker.com/t/com-docker-hyperkit-up-cpu-to-340/13057/73) comment in Docker's forum. Has anyone tried it?  The same thing for me, maybe it helps someone.

**Comment by emil-konstantinovitz** [comment]
I've got the same issue, why is this not reopened?

**Comment by jacoscaz** [comment]
Same issue(s) here. I've replaced my containers with a few `bash` scripts that automate the management of the development environment in question, completely removing Docker from my workflow. A significant loss in terms of development experience but my machine feels _new_.

There seems to be different factors at play but I'm having a hard time figuring out which is which. On one side, it sounds like there's an intrinsic limitation in the way memory management is done throughout the `Mac OS -> HyperKit -> Kernel` chain. On the other side, such a difference in _overall_ performance seems to transcend what I've read about memory.

Other relevant issues I've checked:

https://github.com/moby/hyperkit/issues/231
https://github.com/docker/for-mac/issues/3232

**Comment by radeckim** [comment]
Has anyone resolved the issue? That's so frustrating :(

**Comment by Jabster28** [comment]
Atom crashes after docker starts using >3GB memory. Any hotfixes?

**Comment by zerbinidamata** [comment]
Is there any fix to this?

**Comment by davidcpell** [comment]
Noticed Docker was using 3.9 GB with no containers running 🤔 and eventually found this thread. Stopped Docker and started it again...still using 2.9 GB 🤔 

![image](https://user-images.githubusercontent.com/1223820/61491859-25672180-a97e-11e9-805b-1a94e6c70cb0.png)

**Comment by Jessime** [comment]
@davidcpell Exactly the same thing for me. Saw docker at 3.91 GB, and started reading.

**Comment by vlad-traceme** [comment]
bump!

**Comment by strobox** [comment]
Same here. Very strange

**Comment by Nick-Omen** [comment]
same, more swap I give to Docker, more memory is use

**Comment by schmidtsonian** [comment]
Same :(

**Comment by MattRyder** [comment]
This is one of them issues that you slap a `wontfix` on, and then a competitor comes along that doesn't do this, and you suddenly wonder what the hell happened to your market share.

I've stopped docker running at startup, I suppose I'll take the time hit of starting it up, rather than the performance hit on memory, which I'm already being chewed out on by Android Studio.

**Comment by nsuthison** [comment]
I also have the same problem after fresh restart.

![image](https://user-images.githubusercontent.com/11088738/65372405-0ea0bc00-dc9a-11e9-91e6-9b6d6b6740b7.png)

Just want to post it here in a hope to get attention from someone in docker's development team.

**Comment by oygen87** [comment]
i have 3GB+ memory usage on com.docker.hyperkit without any containers running on MacOS Mojave macbook pro

**Comment by marsbible** [comment]
I have 5GB memory usage  after some times running 2 lightweight container.

**Comment by maximelafarie** [comment]
Holà! Any news? It is more than 3 years this issue is opened and nothing is done by the team to deal with what look like one of the most recurrent and blocking issue on this project... But I guess you accommodates with this situation...

![image](https://aws1.discourse-cdn.com/business4/uploads/scopely/original/3X/f/c/fcbf2adae37b9de0ade406ec0130ae0de8cc1d3d.gif)

Here's my config:
![image](https://user-images.githubusercontent.com/5319267/65515348-6a0fbb80-dedf-11e9-9554-68d0d894b740.png)

And here's the activity monitor:
![image](https://user-images.githubusercontent.com/5319267/65515543-c4108100-dedf-11e9-9d51-219c42cd0ffb.png)

How the heck a defined limit can be overpassed at this point? 😑

**Comment by mabdelhamid** [comment]
How could this issue be closed ? i haven't seen any workaround here ..

Anyway, i've got the same problem, Mojave + docker limited to 4cpus and 2GB memory but taking almost 4GB when no container is running

It just kills my mac's performance and even battery duration when i run some lightweight containers

**Comment by daudn** [comment]
<img width="803" alt="Screenshot 2019-10-10 at 10 47 31" src="https://user-images.githubusercontent.com/48911579/66558441-733d9100-eb4b-11e9-8018-d157806ae9e3.png">
I'll just leave this here. Built an image, memory wasn't cleaned up. It takes a good few minutes to restart and bring the memory usage down.

**Comment by jonassvalin** [comment]
Just adding a +1 here, I see the same issue as everyone else in this thread

**Comment by devashish-patel** [comment]
+1

**Comment by kyprifog** [comment]
Ive been watching this issue for 2 years.   How can it go so long without being looked at by someone?  I still have this issue sometimes and just have to kill docker when I'm not using it

**Comment by madskonradsen** [comment]
@kyprifog Because it, as stated by ijc in the first comment, is working as intended. I agree that it's pretty bad. The issue stems from hyperkit, and Macs aren't usually used as production-servers, so they can't be bothered to make it any better then it is right now.

**Comment by ajpinedam** [comment]
+1

**Comment by nicolapalavecino** [comment]
+1

**Comment by jacoscaz** [comment]
Personally I've migrated to maintaining dedicated start/stop scripts for most of the projects I work on. The majority of the containers I used to deal with were reasonably easy to replace with `brew`-based dependency management and `pm2`-based process management. `pm2`'s `save`/`resurrect` commands are particularly helpful in this regard. Maintaining these scripts takes time but, particularly for local development environments, not having to deal with Docker's memory issues multiple times a day makes it worth it. Has anyone looked into alternatives to Docker? A few colleagues of mine have migrated to linux-based work machines, others have switched to running Docker inside a Linux VM (which is also what Docker for Mac does internally). Is there a different container runtime that runs on Mac OS natively and is compatible with Docker containers and the Docker registry?

**Comment by kubal5003** [comment]
It seems that this issue shows the attitude of the Docker team towards one of the most critical issues ever. Instead of working on a solution - even if that involves hyperkit & possibly even Apple itself it the issue goes deeper to the OS level - what does Docker team do? Absolutely nothing. Problem is unresolved for four years now.

**Comment by bicepjai** [comment]
Please reopen this issue.

**Comment by vikaschoudhary16** [comment]
open the issue please

**Comment by vicky-holcomb** [comment]
+1 also seeing this issue.  After a fresh restart, not running any containers, docker is using 4.36GB of memory.

**Comment by sqanet** [comment]
I installed latest Docker desktop on Mac - Catalina - I always quit docker if I am not using it, otherwise about 3GB ram is consumed even with no containers

**Comment by Roxiun** [comment]
+1 Please Reopen the issue

**Comment by karlwbrown** [comment]
+1 would love if the team would focus on a fix; I'm seeing this on OSX mojave. 3.5GB memory and 5.6% cpu, and I'm just running a single tiny container that is mostly sleeping.

**Comment by gaisho** [comment]
Same.  Please reopen this issue.  I'm running at 3.7GB memory on a simple Wordpress application.

**Comment by kaueburiti** [comment]
+1 Please Reopen the issue, it's a lot of memory ):

**Comment by hanan-vian** [comment]
+1 please solve this

**Comment by dmartlabs** [comment]
+1 Please reopen this issue.

**Comment by SeanTurner026** [comment]
No containers running and 2.43gb of memory consumed.

osx 10.14.6
docker desktop 2.2.0.5

**Comment by astrostl** [comment]
TLDR: Docker Desktop for Mac initially consumes so much memory because it is running a VM in the background [1], and it appears to grow (but actually doesn't) because of an apparent bug in memory reporting in Mac OS X itself [2]

[1] Docker Desktop for Mac runs a Linux VM via https://github.com/moby/hyperkit , as documented at https://docs.docker.com/docker-for-mac/docker-toolbox/ . This is how you can run Linux containers on Mac, it is passing them off to the daemon running there. That VM alone consumes 2GB of memory by default, although it can be pulled back to 1GB in preferences.

[2] https://docs.google.com/document/d/17ZiQC1Tp9iH320K-uqVLyiJmk4DHJ3c4zgQetJiKYQM/edit has extensive documentation and testing. Critically: "The system is able to recover memory from Docker Desktop and give it to other processes if the system is under memory pressure. This is not reflected in the headline memory figure in the Activity Monitor, but can be seen by looking at the Real Memory." and "Since the qemu and hyperkit processes show the same double-accounting problem in Activity Monitor despite them being completely separate codebases, we conclude the bug must be in macOS itself, probably the Hypervisor.framework. We have reported the bug to Apple as bug number 48714544."

**Comment by benspaul** [comment]
please fix help

**Comment by eXamadeus** [comment]
> TLDR: Docker Desktop for Mac initially consumes so much memory because it is running a VM in the background [1], and it appears to hold everything it ever consumes (but actually doesn't) because of an apparent bug in memory reporting in Mac OS X itself [2]
> 
> [1] Docker Desktop for Mac runs a Linux VM via https://github.com/moby/hyperkit , as documented at https://docs.docker.com/docker-for-mac/docker-toolbox/ . This is how you can run Linux containers on Mac, it is passing them off to the daemon running there. That VM alone consumes 2GB of memory by default, although it can be pulled back to 1GB in preferences.
> 
> [2] https://docs.google.com/document/d/17ZiQC1Tp9iH320K-uqVLyiJmk4DHJ3c4zgQetJiKYQM/edit has extensive documentation and testing. Critically: "The system is able to recover memory from Docker Desktop and give it to other processes if the system is under memory pressure. This is not reflected in the headline memory figure in the Activity Monitor, but can be seen by looking at the Real Memory." and "Since the qemu and hyperkit processes show the same double-accounting problem in Activity Monitor despite them being completely separate codebases, we conclude the bug must be in macOS itself, probably the Hypervisor.framework. We have reported the bug to Apple as bug number 48714544."

Sigh, this "conclusion" again fails to address the concerns I posted here: https://github.com/docker/for-mac/issues/178#issuecomment-401432885.

THE RESERVED MEMORY IS STILL FAR TOO HIGH FOR AN IDLE PROCESS.  It never dips below 1.6 GB in my testing and this has NOTHING to do with the double reporting bug.  I'm not using "Activity Monitor", so that's a pointless argument for the team.  I'm also aware of the distinctions between Virtual Memory and Reserved Memory, so a bug in VIRT reporting is great to know about, but I could care less as it isn't relevant.  In my examples, RES memory is right there for show and it doesn't go down AT ALL from first boot.  Watch the GIF again.

This "Virtual memory is double reported" is a distraction.  This is NOT RELATED to the issue I'm reporting.  We all know Virtual Memory doesn't matter here.  But the reserved memory not being freed up?  That's a problem.

Honestly, I need to open a new issue with a link to this thread.  Since the issue I'm seeing isn't a memory leak.  It's a lack of freeing memory when pressure is applied...

**Comment by astrostl** [comment]
Are you considering the fact that a Linux VM with a minimum of 1GB of memory hard-allocated to it is running? That's the [1] part of the message.

**Comment by atlanssia** [comment]
+1 Please Reopen the issue

**Comment by zafercavdar** [comment]
No container is running and 11GB ram is allocated by Docker. 
<img width="649" alt="Screen Shot 2020-06-18 at 15 21 38" src="https://user-images.githubusercontent.com/7279205/85019412-7d526d80-b177-11ea-8c0b-002bb5dbae27.png">

**Comment by mohitt008** [comment]
I'm facing the same issue.

**Comment by fredsherburne-splunk** [comment]
I am facing the same issue. 2 GB memory usage right after starting up my machine.

**Comment by deboracosilveira** [comment]
I also facing this issue, almost 3GB here!

**Comment by georgezim85** [comment]
Try to use just specific project directories (for volumes) in Resources -> File Sharing.
I noticed that if I put my entire user home directory, it takes much more memory.

**Comment by shahrukh-alizai** [comment]
OS: macOS Catalina 10.15.5
Docker: 2.3.0.3 (45519)
Channel: Stable
Engine: 19.03.8

_No containers running. Swap size 512 MB._

**Memory usage: 8.33 GB**

Any solution?

![image](https://user-images.githubusercontent.com/5871892/86343773-d7bcf500-bc72-11ea-9fee-7880b77f5830.png)

**Comment by yogeshmurugesan** [comment]
I also face the same issue, No containers are running
Still consumes more then 3 GB of memory (`Docker version 19.03.8, build afacb8b`)

**Comment by eXamadeus** [comment]
> Are you considering the fact that a Linux VM with a minimum of 1GB of memory (and a current default of 2GB) hard-allocated to it is running? That's the [1] part of the message.

Actually, it seems I missed that part when I was originally replying to your comment.  That would explain the constant reserved memory readings I have seen.  Uggh, that's a lot of overhead to have running constantly.  It would be nice if the docker daemon was smart enough to spin down this VM when there are no containers running.  Sure it would result in slower startups, but that would be a nice configuration option.  Something like "memory conscious-mode" or something that you can opt into that will result in slightly longer spin ups for the first running container you launch because it needs to boot the hyperkit.  It could even have a configurable decay time before the hyperkit gets spun back down (once the last running container is stopped).

Anyway just a thought.  It certainly would make leaving the docker daemon on in the background a lot easier for me.
